pprof/parser.go

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package pprof
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type readState int
+
+const (
+	ignore readState = iota
+	samplesHeader
+	samples
+	locations
+	mappings
+)
+
+// funcID is the ID of a given Location in the pprof raw output.
+type funcID int
+
+type rawParser struct {
+	// err is the first error encountered by the parser.
+	err error
+
+	state    readState
+	funcName map[funcID]string
+	records  []*stackRecord
+}
+
+// ParseRaw parses the raw pprof output and returns call stacks.
+func ParseRaw(input []byte) ([]byte, error) {
+	parser := newRawParser()
+	if err := parser.parse(input); err != nil {
+		return nil, err
+	}
+
+	// TODO(prashantv): Refactor interfaces so we use streams.
+	buf := &bytes.Buffer{}
+	if err := parser.print(buf); err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+func newRawParser() *rawParser {
+	return &rawParser{
+		funcName: make(map[funcID]string),
+	}
+}
+
+func (p *rawParser) parse(input []byte) error {
+	reader := bufio.NewReader(bytes.NewReader(input))
+
+	for {
+		line, err := reader.ReadString('\n')
+		if err != nil {
+			if err == io.EOF {
+				if p.state < locations {
+					p.setError(fmt.Errorf("parser ended before processing locations, state: %v", p.state))
+				}
+				break
+			}
+			return err
+		}
+
+		p.processLine(strings.TrimSpace(line))
+	}
+
+	return p.err
+}
+
+func (p *rawParser) setError(err error) {
+	if p.err != nil {
+		return
+	}
+	p.err = err
+}
+
+func (p *rawParser) processLine(line string) {
```

Comment:
Related to the above comment about `err` being a field of `rawParser`, I use error as a value, similar to `errWriter` here:
https://blog.golang.org/errors-are-values

It's a pattern for not constantly checking `if err != nil`